torch.library.register_fake now accepts more types

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124299
* #124200
* #124199
* #124134
* #124180
* #124092
* #124089
* #124071
* __->__ #124066
* #124065
* #124064
* #123937

We allow it to accept:
- a string with the op name
- an opoverload
- a new-style custom op

If any of these are referring to a new-style custom op (created with the
custom_op decorator), then we dispatch to CustomOpDef.register_fake.
Otherwise, we do what we previously did.

Test Plan:
- new tests

zou3519 committed:
[WIP] torch.library.register_fake now accepts more

[ghstack-poisoned]

zou3519 committed:
Update on "torch.library.register_fake now accepts more"




[ghstack-poisoned]

zou3519 committed:
Update on "torch.library.register_fake now accepts more types"


We allow it to accept:
- a string with the op name
- an opoverload
- a new-style custom op

If any of these are referring to a new-style custom op (created with the
custom_op decorator), then we dispatch to CustomOpDef.register_fake.
Otherwise, we do what we previously did.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "torch.library.register_fake now accepts more types"


We allow it to accept:
- a string with the op name
- an opoverload
- a new-style custom op

If any of these are referring to a new-style custom op (created with the
custom_op decorator), then we dispatch to CustomOpDef.register_fake.
Otherwise, we do what we previously did.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "torch.library.register_fake now accepts more types"


We allow it to accept:
- a string with the op name
- an opoverload
- a new-style custom op

If any of these are referring to a new-style custom op (created with the
custom_op decorator), then we dispatch to CustomOpDef.register_fake.
Otherwise, we do what we previously did.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "torch.library.register_fake now accepts more types"


We allow it to accept:
- a string with the op name
- an opoverload
- a new-style custom op

If any of these are referring to a new-style custom op (created with the
custom_op decorator), then we dispatch to CustomOpDef.register_fake.
Otherwise, we do what we previously did.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "torch.library.register_fake now accepts more types"


We allow it to accept:
- a string with the op name
- an opoverload
- a new-style custom op

If any of these are referring to a new-style custom op (created with the
custom_op decorator), then we dispatch to CustomOpDef.register_fake.
Otherwise, we do what we previously did.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "torch.library.register_fake now accepts more types"


We allow it to accept:
- a string with the op name
- an opoverload
- a new-style custom op

If any of these are referring to a new-style custom op (created with the
custom_op decorator), then we dispatch to CustomOpDef.register_fake.
Otherwise, we do what we previously did.

Test Plan:
- new tests

[ghstack-poisoned]

albanD commented on torch/_library/custom_ops.py:
```diff
@@ -441,6 +442,7 @@ def __call__(self, *args, **kwargs):
 
 
 OPDEF_TO_LIB: Dict[str, "library.Library"] = {}
+OPDEFS: Dict[str, CustomOpDef] = {}
```
Should this be a weak value dict?

zou3519 replied:
yup

albanD commented on torch/_library/custom_ops.py:
```diff
@@ -468,3 +470,11 @@ def check(arg):
         yield from check(kwarg)
+
+
+def _maybe_get_opdef(qualname):
```
qualname sounds like it should be a str object which is not true given the below...

albanD replied:
Also it's used only in one place, should we just move it to that other file?

zou3519 replied:
> qualname sounds like it should be a str object which is not true given the below...

Ya

> Also it's used only in one place, should we just move it to that other file?

I'm trying to not pollute library.py too much

zou3519 committed:
Update on "torch.library.register_fake now accepts more types"


We allow it to accept:
- a string with the op name
- an opoverload
- a new-style custom op

If any of these are referring to a new-style custom op (created with the
custom_op decorator), then we dispatch to CustomOpDef.register_fake.
Otherwise, we do what we previously did.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "torch.library.register_fake now accepts more types"


We allow it to accept:
- a string with the op name
- an opoverload
- a new-style custom op

If any of these are referring to a new-style custom op (created with the
custom_op decorator), then we dispatch to CustomOpDef.register_fake.
Otherwise, we do what we previously did.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "torch.library.register_fake now accepts more types"


We allow it to accept:
- a string with the op name
- an opoverload
- a new-style custom op

If any of these are referring to a new-style custom op (created with the
custom_op decorator), then we dispatch to CustomOpDef.register_fake.
Otherwise, we do what we previously did.

Test Plan:
- new tests

[ghstack-poisoned]

albanD reviewed (approved):
Thanks for the update!

zou3519 committed:
Update on "torch.library.register_fake now accepts more types"


We allow it to accept:
- a string with the op name
- an opoverload
- a new-style custom op

If any of these are referring to a new-style custom op (created with the
custom_op decorator), then we dispatch to CustomOpDef.register_fake.
Otherwise, we do what we previously did.

Test Plan:
- new tests

[ghstack-poisoned]
